[lldb] Updated lldb-server to spawn the child process and share socket

slydiman opened the pull request:
`lldb-server platform --server` works on Windows now w/o multithreading. The rest functionality remains unchanged.

Fixes #90923, fixes #56346.

This is the part 1 of the replacement of #100670.

In the part 2 I plan to switch `lldb-server gdbserver` to use `--fd` and listen a common gdb port for all gdbserver connections. Then we can remove gdb port mapping to fiх #97537.

slydiman committed:
[lldb] Updated lldb-server to spawn the child process and share socket on Windows

`lldb-server platform --server` works on Windows now w/o multithreading. The rest functionality remains unchanged.

Depends on #101383.

Fixes #90923, fixes #56346.

This is the part 1 of the replacement of #100670.

In the part 2 I plan to switch `lldb-server gdbserver` to use `--fd` and listen a common gdb port for all gdbserver connections. Then we can remove gdb port mapping to fix #97537.

slydiman commented:
Note GDBRemoteCommunicationServerPlatform::DebugserverProcessReaped() may cause a crash if this callback will be called from the child monitor thread after destroing the instance of GDBRemoteCommunicationServerPlatform. It is impossible to control this thread anyway.
Hope I will fix it in the part 2 removing m_port_map. m_spawned_pids may be static since we have one GDBRemoteCommunicationServerPlatform instance per process.

DavidSpickett commented on lldb/source/Host/windows/PipeWindows.cpp:
```diff
@@ -283,54 +293,94 @@ Status PipeWindows::ReadWithTimeout(void *buf, size_t size,
+  if (!result) {
+    if (GetLastError() != ERROR_IO_PENDING)
+      return Status(::GetLastError(), eErrorTypeWin32);
+    else {
```
Don't need the else given that the line above is a return.

slydiman replied:
Thanks. Updated.

DavidSpickett commented on lldb/source/Host/windows/PipeWindows.cpp:
```diff
@@ -283,54 +293,94 @@ Status PipeWindows::ReadWithTimeout(void *buf, size_t size,
-      BOOL cancel_result = CancelIoEx(m_read, &m_read_overlapped);
-      if (!cancel_result && GetLastError() == ERROR_NOT_FOUND)
-        failed = false;
+  if (!result) {
```
Could do an early return here `if (result) {`. You'll have 2 copies of:
```
bytes_read = sys_bytes_read;
return Status();
```
But it could be easier to read.

slydiman replied:
Thanks. Updated.

DavidSpickett commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -159,6 +159,40 @@ GDBRemoteCommunicationServerPlatform::GDBRemoteCommunicationServerPlatform(
+                                   "", // hostname
+                                   pid, port, socket_name);
+    if (error.Success())
+      SetPendingGdbServer(pid, *port, socket_name);
```
What happens with the implicit `else` part of this, does `GetPacketAndSendResponse` somehow fail immediately?

slydiman replied:
This code is just moved from lldb-platform.cpp 
GetPacketAndSendResponse() ignores `error` value. GetPacketAndSendResponse() must set the flag `quit` or `interrupt` and update `error`.

DavidSpickett commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -159,6 +159,40 @@ GDBRemoteCommunicationServerPlatform::GDBRemoteCommunicationServerPlatform(
+    LLDB_LOGF(log,
+              "GDBRemoteCommunicationServerPlatform::%s() "
+              "GetPacketAndSendResponse: %s",
+              __FUNCTION__, error.AsCString());
```
I would've thought that `__FUNCTION__` gave you the function name here but maybe not. Can you just try making it `if (true)` and confirm what it prints?

DavidSpickett commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerPlatform.cpp:
```diff
@@ -159,6 +159,40 @@ GDBRemoteCommunicationServerPlatform::GDBRemoteCommunicationServerPlatform(
 GDBRemoteCommunicationServerPlatform::~GDBRemoteCommunicationServerPlatform() =
     default;
 
+void GDBRemoteCommunicationServerPlatform::Proc(
```
If we're changing the name could we have a better name than `Proc`, or if `Proc` is just a term I am unfamiliar with, a comment in the header to explain what this function does?

slydiman replied:
Note `void SetInferiorArguments(const lldb_private::Args &args);` is a missing function. It is a bug. I have just removed it and added Proc(). There is no any connection between these functions. I have just moved a common code from lldb-platform to GDBRemoteCommunicationServerPlatform::Proc(). We can rename it to ClientHandle() or such and even kept inside lldb-platform.cpp

DavidSpickett commented on lldb/tools/lldb-server/LLDBServerUtilities.cpp:
```diff
@@ -27,11 +27,13 @@ class TestLogHandler : public LogHandler {
       : m_stream_sp(stream_sp) {}
 
   void Emit(llvm::StringRef message) override {
+    std::lock_guard<std::mutex> guard(m_mutex);
```
A mutex is now required because of the sharing of the socket I assume?

Could you add a comment on the member variable to say that as well.

slydiman replied:
Probably I can separate this change to another patch. It is not related to the socket sharing. It is the fix of an old bug. 
Host::LaunchProcess() requires to SetMonitorProcessCallback. This callback is called from the child process monitor thread. We cannot control this thread. lldb-server may crash if there is a logging around this callback because TestLogHandler is not thread safe.
I've kept this code here for now because I faced this issue only debugging this patch with 100 simultaneous child processes. It is hard to reproduce it other way. Note also [this comment](https://github.com/llvm/llvm-project/pull/101283#issuecomment-2259628620).

DavidSpickett commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +120,218 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+  if (num_bytes != sizeof(protocol_info)) {
+    LLDB_LOGF(log,
+              "lldb-platform child: "
+              "socket_pipe(0x%x).ReadWithTimeout(WSAPROTOCOL_INFO) failed: "
```
This is missing accept_fd for the first %x I think.

slydiman replied:
Thanks. Updated.

DavidSpickett reviewed (commented):
What, if any, are the changes to how you would run lldb-server on Windows? I see a new argument here but not sure if that's only meant for lldb-server to pass to itself when it starts a child process.

DavidSpickett commented:
What exactly does `lldb-server --server` mean? Because at least on Linux I see:
```
$ ./bin/lldb-server
Usage:
  ./bin/lldb-server v[ersion]
  ./bin/lldb-server g[dbserver] [options]
  ./bin/lldb-server p[latform] [options]
Invoke subcommand for additional help
```
Maybe you just have a typo or `--server` is a shortcut for one of those.

DavidSpickett commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +120,218 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+static GDBRemoteCommunicationServerPlatform::PortMap gdbserver_portmap;
+static std::mutex gdbserver_portmap_mutex;
+
+#if defined(_WIN32)
```
Do you think we will always have these large __WIN32 blocks, or if we go forward with your plan, will some of that become generic code once again?

slydiman replied:
We can move this code to some windows specific directory. But it is windows specific anyway. Note it contains a lot of `lldb-server platform` specific parameters. We can just pass the socket FD via the command line on non-Windows OS. `lldb-server gdbserver` already supports it with the parameter `--fd`. But it does not work on Windows. acceprt_fd here is a pipe FD which is used to transfer the structure WSAPROTOCOL_INFO. We need to know the child pid to initialize WSAPROTOCOL_INFO. Finally the child process is able to initialize the accepted socket FROM_PROTOCOL_INFO.

labath reviewed (commented):
I think we should slow down a bit.

For consistency, and to minimize the number of ifdefs, I believe windows and non-windows paths should use as similar approaches as possible. That means I think the end state should be that the posix path also uses a fork+exec approach.

For me, it also seems natural that this should be the first step. While doing that, we can make sure to create abstractions that can be used to implement the windows behavior as well. I'm imagining some sort of a function or a class, which takes the target binary, its arguments, and a socket, and launches that binary with that socket. (we'll probably also need a similar function/class on the other end to receive that socket).

The function doesn't have to be terribly generic, but it ideally should encompass all of platform specific behavior, so that the surrounding code does not need to use ifdefs. (But, ideally, it should at least be slightly generic, so that one can write a unit test for this functionality).

(BTW, I'm OOO for the rest of the week, so I won't be able to review this in detail. Upon cursory inspection, the general idea makes sense, but we really need to figure out how to encapsulate this. I'm not sure if we really need the WriteWithTimeout functionality. The reason we don't have it so far is that pipes have a buffer, and unless we're writing a huge chunk of data, the write will never block anyway. There's nothing wrong with it in principle though, and if you do want to have it, it'd be best if it comes with in its own patch and with a test.)

slydiman commented:
> What exactly does `lldb-server --server` mean?

Sorry. I have updated the description with  `lldb-server platform --server`.

slydiman commented:
> What, if any, are the changes to how you would run lldb-server on Windows? I see a new argument here but not sure if that's only meant for lldb-server to pass to itself when it starts a child process.

The explanation [is here](https://github.com/llvm/llvm-project/pull/101283#discussion_r1698289698). This new argument is the fork() workaround for Windows. We can keep this new argument for non-Windows OS, but it is redundant because fork() is enough.

slydiman commented:
@labath 

> For consistency, and to minimize the number of ifdefs, I believe windows and non-windows paths should use as similar approaches as possible. That means I think the end state should be that the posix path also uses a fork+exec approach.

fork+exec on non-Windows OS is redundant. It simply does not make sence. We can use multithreading or CreateProcess on Windows. It seems [the multithreading solution](https://github.com/llvm/llvm-project/pull/100670) faced a lot of issues and system's bugs on Linux. So CreateProcess is the only way. But it is the fork() workaround for Windows.

> For me, it also seems natural that this should be the first step. While doing that, we can make sure to create abstractions that can be used to implement the windows behavior as well. I'm imagining some sort of a function or a class, which takes the target binary, its arguments, and a socket, and launches that binary with that socket. (we'll probably also need a similar function/class on the other end to receive that socket).

Look at GDBRemoteCommunication::StartDebugserverProcess(). I planed to extend it to use `pass_comm_fd` of Windows in the part 2. It is hard to create abstractions because of a lot of parameters. We can move most of SpawnProcessChild() code to `ConnectionFileDescriptor::ConnectFD` in ConnectionFileDescriptorPosix.cpp BTW, I don't like #ifdef _WIN32 in *Posix.cpp files. But currently the common ConnectionFileDescriptor.h contains `#include "lldb/Host/posix/ConnectionFileDescriptorPosix.h"`

> I'm not sure if we really need the WriteWithTimeout functionality. The reason we don't have it so far is that pipes have a buffer, and unless we're writing a huge chunk of data, the write will never block anyway. There's nothing wrong with it in principle though, and if you do want to have it, it'd be best if it comes with in its own patch and with a test.)

Before this patch the pipe read/write 1 byte in ConnectionFileDescriptor::CommandPipe. The maximal size was 6 bytes (the port value) in GDBRemoteCommunication::StartDebugserverProcess() and it is only the place where ReadWithTimeout() is used. But WSAPROTOCOL_INFO is big enough (628 bytes). Usually the pipe buffer is 4K. But we need an ability to avoid hangs anyway.
I agree that all pipe related changes may be moved to another patch. And we need to add WriteWithTimeout() to PipePosix too. Note there is not any test for Pipe now. And I have no idea how to test it, especially async version. Running `lldb-server platform --server` on Windows is the best test. It must be tested with a high load with multiple connections. Otherwise any pipe tests are useles.

labath commented:
> @labath
> 
> > For consistency, and to minimize the number of ifdefs, I believe windows and non-windows paths should use as similar approaches as possible. That means I think the end state should be that the posix path also uses a fork+exec approach.
> 
> fork+exec on non-Windows OS is redundant. It simply does not make sence. We can use multithreading or CreateProcess on Windows. It seems [the multithreading solution](https://github.com/llvm/llvm-project/pull/100670) faced a lot of issues and system's bugs on Linux. So CreateProcess is the only way. But it is the fork() workaround for Windows.

This is where I'll have to disagree with you. The exec following the fork is not "redundant" on linux. A naked fork is an extremely dangerous thing in todays (multithreaded) applications. I'm absolutely certain that the improper use of fork with threads (and the FD leaks that go along with it) is the cause of at least one of the issues (ETXTBSY) you refer to as "system bugs", and fairly sure it's also the cause of the other one (the pipe workaround) as well. In a multithreaded application (which lldb-platform isn't right now, but more of an accident than design, and which you've tried to make extremely multithreaded) the only safe use of fork is if it is immediately followed by an exec, and this has to be done in an extremely careful way (we try to do that, and we still failed).

But even if it weren't for all of this, I'd still think it makes sense to start a new process, just for symmetry with windows. That's the nature of portable programming.

> 
> > For me, it also seems natural that this should be the first step. While doing that, we can make sure to create abstractions that can be used to implement the windows behavior as well. I'm imagining some sort of a function or a class, which takes the target binary, its arguments, and a socket, and launches that binary with that socket. (we'll probably also need a similar function/class on the other end to receive that socket).
> 
> Look at GDBRemoteCommunication::StartDebugserverProcess(). I planed to extend it to use `pass_comm_fd` of Windows in the part 2. It is hard to create abstractions because of a lot of parameters. We can move most of SpawnProcessChild() code to `ConnectionFileDescriptor::ConnectFD` in ConnectionFileDescriptorPosix.cpp BTW, I don't like #ifdef _WIN32 in *Posix.cpp files. But currently the common ConnectionFileDescriptor.h contains just `#include "lldb/Host/posix/ConnectionFileDescriptorPosix.h"`

I didn't say it was going to be easy. :) I can look at this closer when I get back next week, but I definitely think we can and should come up with a solution involving less ifdefs.

> 
> > I'm not sure if we really need the WriteWithTimeout functionality. The reason we don't have it so far is that pipes have a buffer, and unless we're writing a huge chunk of data, the write will never block anyway. There's nothing wrong with it in principle though, and if you do want to have it, it'd be best if it comes with in its own patch and with a test.)
> 
> Before this patch the pipe read/write 1 byte in ConnectionFileDescriptor::CommandPipe. The maximal size was 6 bytes (the port value) in GDBRemoteCommunication::StartDebugserverProcess() and it is only the place where ReadWithTimeout() is used. But WSAPROTOCOL_INFO is big enough (628 bytes). Usually the pipe buffer is 4K. 

628 is still less that 4k. Can the buffer be smaller than that?

> But we need an ability to avoid hangs anyway. I agree that all pipe related changes may be moved to another patch. And we need to add WriteWithTimeout() to PipePosix too. Note there is no any test for Pipe now. And I have no idea how to test it, especially async version. Running `lldb-server platform --server` on Windows is the best test. It must be tested with a high load with multiple connections. Otherwise any pipe tests are useles.

We have one and a half pipe tests (one of them claims to be broken on windows). I wrote them when fixing some Pipe bugs. We should definitely have more.

I wholeheartedly disagree with the assertion that the test is useless unless it uses multiple threads and high load. Even the most simple test serves as documents the correct/expected API usage and proves that the functionality works at least in the simple cases. For something like `WriteWithTimeout` I'd probably create single threaded test which does something like this:
- write to the pipe until it is full
- try to write with a larger timeout and check that it actually times out
- read from the pipe and check that you got back what was written

That will cover most of the codepaths and obvious ways it can go wrong (e.g. during refactoring). Then, if you wanted to be fancy, you could do things like start a write with a long timeout and then try to read from the pipe, and see that causes the write to complete, etc.; but I wouldn't expect everyone to write those all the time.

slydiman commented:
@labath 
> one of the issues (ETXTBSY) you refer to as "system bugs"

I meant https://bugzilla.kernel.org/show_bug.cgi?id=546 and [this discussion](https://github.com/llvm/llvm-project/pull/100670/files#r1693201568). I'm sad that you don't believe. I concluded that no one uses the real multithreading on Linux. Someone tried and faced unsolvable problems. Therefore, #100670 is doomed.

slydiman commented:
I have moved all Pipe related changes to #101383.

labath commented:
> @labath
> 
> > This is where I'll have to disagree with you. The exec following the fork is not "redundant" on linux. A naked fork is an extremely dangerous thing in todays (multithreaded) applications. I'm absolutely certain that the improper use of fork with threads (and the FD leaks that go along with it) is the cause of at least one of the issues (ETXTBSY) you refer to as "system bugs", and fairly sure it's also the cause of the other one (the pipe workaround) as well. In a multithreaded application (which lldb-platform isn't right now, but more of an accident than design, and which you've tried to make extremely multithreaded) the only safe use of fork is if it is immediately followed by an exec, and this has to be done in an extremely careful way (we try to do that, and we still failed).
> 
> I meant https://bugzilla.kernel.org/show_bug.cgi?id=546 and [this discussion](https://github.com/llvm/llvm-project/pull/100670/files#r1693201568). I'm sad that you don't believe.

I know what you're referring to. However, that's a statement that comes with a very heavy burden of proof, one that you simply have not provided. The number of users of linux kernel exceeds the users of lldb by many orders of magnitude, so it's very unlikely that the problem is on their side. It's not that it's impossible -- I personally have tracked down lldb problems down to bugs in our dependencies -- from the libstdc++ to the kernel. However, this just doesn't look like one of those cases.

> I concluded that no one uses the real multithreading on Linux.

That's another very bold statement. I am not sure what you meant by "real multithreading", but I know for a fact that some Linux applications run thousands of threads in one process without any issues. All of them are very careful about how they use fork though.

> Someone tried and faced unsolvable problems. Therefore, #100670 is doomed.
> 
> > But even if it weren't for all of this, I'd still think it makes sense to start a new process, just for symmetry with windows. That's the nature of portable programming.
> 
> Not all code may be portable. It is better to use threads on Windows, but not on Posix. fork() is missing on Windows and starting a new process is a workaround. When we have the accepted socket and everything is ready to just handle the connection on Posix (after fork), you suggest to drop all initialized parameters, start a new process, pass all necessary parameters via the command line, initialize all parameters again, go to the same point and finally handle the connection. What is the reason? To be similar how it will work on Windows.

That is exactly what I'm suggesting.

> No, I don't agree.

Well, at least we can agree to disagree. ;)

> You are worried about some FD. But the main process `lldb-server platform` does not launch anything. There are no any FDs which may leak.

It forks. Forking duplicates any FD opened into the parent into the child. Even those marked with CLOEXEC, which is the usual way to guard against FD leakage. If you're single-threaded than this is manageable, but as soon as you have multiple threads (believe it or not, many applications on linux use multiple threads, and a lot of code creates background threads; lldb-server platform code executed here is pretty small, but it still calls into a lot of code that could create those), fork() becomes a ticking time bomb. In a multithreaded application, the only code that can be safely run after a fork, is that which can run in a signal handler. You can't write a server like that.

Using fork to serve multiple connections is a pattern from the seventies. Modern applications don't do it this way. As far as I'm concerned, this code should not have been written this way (and if the author cared about windows, they would have never done it), and I'd support removing this pattern even if it weren't for windows.

> We can't unify socket sharing nicely for Windows and Posix. We can try to combine the common code in the part 2 of this patch when the socket sharing will be added to GDBRemoteCommunication::StartDebugserverProcess() too.

That's nice, but I think we should figure out how to reduce the number of ifdefs in this patch. Porting linux away from fork may not be your concern, but figuring out how to make the code less branchy is. If you can do that without removing forks, I can take it upon myself to handle the rest. I just think it would be easier to do it both together...

slydiman commented:
@labath 
>That's nice, but I think we should figure out how to reduce the number of ifdefs in this patch. Porting linux away from fork may not be your concern, but figuring out how to make the code less branchy is. If you can do that without removing forks, I can take it upon myself to handle the rest. I just think it would be easier to do it both together...

`#ifdef _WIN32` is used here for the following
- declare and initialize the parameter --accept (accept_fd)
- call spawn_process_child() and exit if accept_fd is defined
- call spawn_process_parent() instead of fork()

We can unify spawn_process_child() moving it to ConnectionFileDescriptor::ConnectFD() in ConnectionFileDescriptorPosix.cpp with `#ifdef _WIN32`.

Note I will remove `#ifndef _WIN32` around ::waitpid() in lldb-platform.cpp in the part 2 of this patch.
We still need spawn_process_reaped() here but I will remove it in part 2 too. Any static dummy will be enough.
For now we need spawn_process_reaped() here for `gdbserver_portmap.FreePortForProcess(pid);`.

We can try to move spawn_process_parent() to `lldb/source/Host/windows/` and probably a part to `lldb/source/Host/posix/`.
Where is the best place? HostProcess*.cpp, ProcessLauncher*.cpp?
It is necessary to implement some callback for spawn_process_reaped(), which will be removed in part 2.
It looks like a redundant work now.

Probably it is necessary to define a new type and an invalid value for accept_fd. It is int (socket fd) on Posix and pipe_t (HANDLE) on Windows.
What is the best type name and what is the best place to define it?

I would be happy to keep it as is till part 2 of this patch to minimize a redundant work.

slydiman committed:
[lldb] Updated lldb-server to spawn the child process and share socket

`lldb-server platform --server` works on Windows now w/o multithreading. The rest functionality remains unchanged.

Fixes #90923, fixes #56346.

This is the part 1 of the replacement of #100670.

In the part 2 I plan to switch `lldb-server gdbserver` to use `--fd` and listen a common gdb port for all gdbserver connections. Then we can remove gdb port mapping to fix #97537.

slydiman commented:
I have updated the patch to spawn the child process on non-Windows too and minimized the number of `#ifdef _WIN32`.

labath commented:
I would start looking at this from the other end. Let's identify logical pieces of functionality that can be shared between the various modes of operation (fork/exec/singleserver), and try to refactor to code to fit that. While doing that, I think an important criteria should be whether some of the code is platform-specific. If it is, the function containing that should be as small as possible so that the ifdef doesn't occur in a deeply nested code (deeply nested ifdefs are much harder to reason about than one that occurs at the top level.

The first obvious piece of functionality I see is the one which serves a single connection. This is something we need to execute in all three code paths. It should definitely be a function of its own (I'm going to call it `ServeSingleConnection` here, but there may be a name that better fits local conventions), so that we can just call it, instead of reaching it through a web of compile and runtime conditionals.

The second one is the piece that serves multiple connections (let's call it `ServeManyConnections`) and handles handing them off to subprocesses. This one is tricky because it going to be platform specific. I'm not clear on whether it should be a single function with embedded ifdefs or two implementations of a single function, but I think it should be as small as possible so that the platform specificities really stand out.

I would try to implement it as something like this (where the two functions in the while loop are platform-specific code):
```
void ServeManyConnections(...) {
  while (conn = AcceptNewConnection()) {
    ServeInSubprocess(conn);
    ReapSubprocessesIfNeeded();
  }
}
```

... but I can also image other schemes. (BTW, I don't think you can get rid of the waitpid calls completely, as you need to reap the zombies even if you're not going to do anything with them).

Then I think we can structure the main (or something main-like) function as something like:
```
int main() {
  ...

  if (getting_connection_from_parent)
    return AcceptConnectionFromParent(...); // retrieves the connection and calls ServeSingleConnection

  if (!server_mode) // no --server flag
    return ServeSingleConnection(...)

  return ServeManyConnections(...);
}
```

I know this is handwavy, but that's sort of the goal I think we should aim for. It will get messier once all of the arguments (which I've conveniently ignored here) get added, but feel free to add structs or classes to group things as they make sense. Once we have this, I think it will also be clearer which code (if any) we want to put in some common libraries. Doing everything at once also may not be the best idea -- I can totally see how a preparatory patch which just refactors things without adding windows support would make sense here.

(Also @DavidSpickett, feel free to chime in, I shouldn't be the only one dictating conditions here.:)

labath commented:
We raced updating the PR, so I didn't see the new version. I think the new version is looking better. I'm going to comment in more detail soon.

slydiman commented:
@labath Please look at the updated patch.
ServeSingleConnection() = client_handle()
AcceptConnectionFromParent() = some code + client_handle()
ServeInSubprocess() = spawn_process()

lldb-platform.cpp is simple. Currently we don't see the whole picture before the part 2 of this patch. The socket sharing for `lldb-server gdbserver` from GDBRemoteCommunication::StartDebugserverProcess() is much more complicated. And we need to update the acceptor to listen 2 ports in the same thread. I think we should implement the new functionality first and then we can optimize it.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -44,6 +47,14 @@ using namespace llvm;
 
+#ifdef _WIN32
+typedef pipe_t fd_t;
+#define kInvalidFD (LLDB_INVALID_PIPE)
```
Lets make this a real constant (not a macro).

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -60,6 +71,7 @@ static struct option g_long_options[] = {
     {"max-gdbserver-port", required_argument, nullptr, 'M'},
     {"socket-file", required_argument, nullptr, 'f'},
     {"server", no_argument, &g_server, 1},
+    {"fd", required_argument, nullptr, 'd'},
```
We don't expect this option to be used by the end user (at least not on windows, where the passing algorithm is quite complex), right?

If so, then I think we should drop the short option (the field is `int`, so I believe you can just use number not corresponding to a printable char), and use something longer for the long option -- to reduce the likelyhood of someone using this by mistake. I'd go with something like `--child-platform-fd`..

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +126,175 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+                            const std::string &log_file,
+                            const StringRef log_channels) {
+  Status error;
+  const TCPSocket &tcpSocket =
+      static_cast<const TCPSocket &>(*conn->GetReadObject());
```
Other types of connections can be used here (at least on unixen, anyway). The cast should only be necessary on windows (and even there, the code should probably check the type first).

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +126,175 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+
+#ifdef _WIN32
+  // Transfer WSAPROTOCOL_INFO to the child process.
+  if (socket_pipe.CanRead())
+    socket_pipe.CloseReadFileDescriptor();
+  if (!socket_pipe.CanWrite()) {
```
When can these fail? I would expect these to always hold and not even be worthy of an assert..

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +126,175 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+  ProcessLaunchInfo launch_info;
+
+  fd_t fd;
+#ifdef _WIN32
```
I think these two ifdefs in this function could be extracted to a separate class. I'm imagining something with a constructor which takes the `Connection` object (maybe also the ProcessLaunchInfo), and two methods that:
- return the number that should be passed to the subprocess via arguments (the unix version will just pass the connection fd). `GetSendableFD` ?
- finish the sending after the new process is started (unix version would be a noop). `CompleteSending` ?

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +126,175 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+
+  size_t num_bytes;
+  error = socket_pipe.WriteWithTimeout(&protocol_info, sizeof(protocol_info),
+                                       std::chrono::seconds(2), num_bytes);
```
2 seconds seems a bit on the low side. Since this fits in the pipe buffer, I guess it will never block in practice , but if it did, 2 seconds might not be enough on a loaded machine. I'd put 10s, just to prevent anyone from guessing if this is the source of the flakyness.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -254,6 +445,65 @@ int main_platform(int argc, char *argv[]) {
+    }
+
+    NativeSocket socket;
+#ifdef _WIN32
```
And I think this could be a static method on the class I described above. `CompleteReceiving` ?

labath reviewed (commented):
I actually quite like this. I think we should further merge the platform specific code into one place (which, among other things, will reduce the total number of ifdefs to ~one, and put all of the socket handling code next to each other), but apart from that, I think this is great (all things considered, etc).

labath commented:
Also, if any of this code is not tested (outside of running the entire test suite in the remote mode) we should add a dedicated test for it. A quick search found at least one (test/API/tools/lldb-server/TestGdbRemoteCompletion.py) test which would appear to exercise this, but maybe there's room for more.

slydiman committed:
Added the class SharedSocket.

slydiman commented:
@labath I have added the class SharedSocket. I have implemented all the recommendations.
Note we have the buildbot for cross tests (Windows x64/Linux x64 host and Linux Aarch64 target).
So everything is tested. Hope 1191 API tests run with this lldb-server are enough.

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -44,6 +47,112 @@ using namespace llvm;
+    return Status();
+  }
+
+  Status GetNativeSocket(NativeSocket &socket) {
```
I think this would be better off as a static function -- it's not using any of the members except m_fd, which could just as easily be a function argument, and it's duplicating the Pipe and NativeSocket members as local vars.

I can also imagine with a proper member function version of it, but then I think should reuse the members in the class. For example, by putting most of this work into the constructor, where it will initialize the m_pipe member and use it to receive the NativeSocket. `GetNativeSocket` could then be a plain getter (?)

labath commented on lldb/tools/lldb-server/lldb-platform.cpp:
```diff
@@ -114,6 +224,133 @@ static Status save_socket_id_to_file(const std::string &socket_id,
+  launch_info.SetLaunchInSeparateProcessGroup(false);
+  launch_info.SetMonitorProcessCallback(&spawn_process_reaped);
+
+  // Copy the current environment.
+  // WSASocket(FROM_PROTOCOL_INFO) will fail in the child process
+  // with the error WSAEPROVIDERFAILEDINIT if the SystemRoot is missing
+  // in the environment.
```
FWIW, I don't think this requires any special justification. Environment inheritance is the default behavior when launching processes. I'd just delete this comment.

labath reviewed (commented):
> @labath I have added the class SharedSocket. I have implemented all the recommendations.

Looks mostly fine. Just a couple of notes. David, what do you think?


> Note we have the buildbot for cross tests (Windows x64/Linux x64 host and Linux Aarch64 target).
> So everything is tested. Hope 1191 API tests run with this lldb-server are enough.

That's great, but I don't think this should be the only way this code is tested, as most of it can be tested perfectly well locally. Most of the devs will not be running this bot configuration before they push their patches, and they will not know how to set it up when they get the breakage email, so it's always better to have a first line of defense that everyone can use, and leave the cross testing for catching subtle integration details.

That said, we do have that one test which should exercise most of this code, so I'm sort of fine with this for now, but I still may ask for more tests in the future, for example, when we librarize this code.

slydiman committed:
Made GetNativeSocket() a static function.

slydiman committed:
Removed waitpid(). spawn_process_reaped() does its job.

slydiman committed:
Renamed `fd_t` to `shared_fd_t`.

labath reviewed (approved):
This looks good. Thanks for your patience.
